### PR TITLE
feat(mobile): edit date time & location

### PR DESCRIPTION
**Pull request by shenlong-tanwen**
#### Changes made in the PR

- Initial support for edit location and date and time from timeline as well as asset viewer
- Location can be either entered manually or can be selected from a map

#### Screenshots

| Bottom Sheet | Edit Date Time | Edit Location | Manual Location
| :-: | :-: | :-: | :-: |
Bottom sheet | Edit date | Location Edit Dialog | Manual Location

| Map Picker | Bottom sheet | Asset without location |
| :-: | :-: | :-: |
Map picker | Exif Info Sheet | Asset without location

### CODE REVIEW

File: mobile/lib/modules/activities/providers/activity.provider.dart [hunk 1]
```diff
@@ -95,7 +95,11 @@ class ActivityStatisticsNotifier extends StateNotifier<int> {
   }
 
   Future<void> fetchStatistics() async {
-    state = await _activityService.getStatistics(albumId, assetId: assetId);
+    final count =
```
Comment (fyfrey) [inline mobile/lib/modules/activities/providers/activity.provider.dart:98]: Why make the request at all if not mounted?

File: mobile/lib/utils/selection_handlers.dart [hunk 2]
```diff
@@ -85,3 +90,60 @@ Future<void> handleFavoriteAssets(
...
+    context: context,
+    initialDateTime: initialDate,
+    initialTZ: timeZone,
+    initialTZOffset: offset,
+  );
+  if (dateTime == null) {
+    return;
+  }
+
+  ref.watch(assetServiceProvider).changeDateTime(selection.toList(), dateTime);
```
Comment (fyfrey) [inline mobile/lib/utils/selection_handlers.dart:120]: watch -> read in any non-build function (e.g. callback like here)

**Review by fyfrey — COMMENTED** [review]
Awesome addition to the app!

I've left some small remarks here and there.

The edit buttons on the bottom action sheet have quite a long text compared to the icon width and the other icon buttons. Can we find shorter texts?

File: mobile/lib/extensions/asset_extensions.dart [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import 'package:immich_mobile/shared/models/asset.dart';
+import 'package:timezone/timezone.dart';
+
+extension TZExtension on Asset {
+  (DateTime, Duration) getTZAdjustedTimeAndOffset() {
```
[the remark this answers is not in the document]
Comment (shenlong-tanwen) [reply mobile/lib/extensions/asset_extensions.dart:8]: Added doc comments and test

File: mobile/lib/modules/activities/providers/activity.provider.dart [hunk 4]
```diff
@@ -95,7 +95,11 @@ class ActivityStatisticsNotifier extends StateNotifier<int> {
...
+        await _activityService.getStatistics(albumId, assetId: assetId);
+    if (mounted) {
```
Comment (alextran1502) [inline mobile/lib/modules/activities/providers/activity.provider.dart:100]: Where does this `mounted` variable come from?

File: mobile/lib/modules/activities/providers/activity.provider.dart (hunk 1 above)
Comment (shenlong-tanwen) [reply mobile/lib/modules/activities/providers/activity.provider.dart:98]: We'll send the request right after a user opens an album that is shared with other users. However, if the user immediately navigates back to the library page, the response for the request initiated is assigned to the state. However, the widget would've been unmounted and hence, the provider as well, resulting in an error.

File: mobile/lib/modules/activities/providers/activity.provider.dart (hunk 4 above)
Comment (shenlong-tanwen) [reply mobile/lib/modules/activities/providers/activity.provider.dart:100]: It is part of `StateNotifier`

File: mobile/lib/utils/selection_handlers.dart (hunk 2 above)
Comment (shenlong-tanwen) [reply mobile/lib/utils/selection_handlers.dart]: Updated

File: mobile/lib/utils/selection_handlers.dart [hunk 5]
```diff
@@ -85,3 +90,60 @@ Future<void> handleFavoriteAssets(
...
+  }
+  final location = await showLocationPicker(
+    context: context,
+    initialLatLng: initialLatLng,
+  );
+  if (location == null) {
+    return;
+  }
+
+  ref.watch(assetServiceProvider).changeLocation(selection.toList(), location);
```
[the remark this answers is not in the document]
Comment (shenlong-tanwen) [reply mobile/lib/utils/selection_handlers.dart]: Updated

**Review by alextran1502 — APPROVED** [review]

**Comment by oziee** [comment]
I get an error with this new feature when an item has an absurd date.. was hoping this new feature would allow me to fix this ONE item :)

Screenshot 2023-12-06 at 9 30 05 pm
Screenshot 2023-12-06 at 9 30 21 pm

```
flutter: [SEVERE] [2023-12-06 21:32:25.961626] Catch all error: 'package:flutter/src/material/date_picker.dart': Failed   
assertion: line 198 pos 5: '!initialDate.isAfter(lastDate)': initialDate 236276-12-31 00:00:00.000 must be on or before lastDate  
 2023-12-06 00:00:00.000. - 'package:flutter/src/material/date_picker.dart': Failed assertion: line 198 pos 5:   
'!initialDate.isAfter(lastDate)': initialDate 236276-12-31 00:00:00.000 must be on or before lastDate 2023-12-06   
00:00:00.000.  

```

**Comment by shenlong-tanwen** [comment]
> can something in code be added to clear the date for wacky items so they can be altered

Handled it in #5522. Thank you for the notice.

**Pull request merged**